### PR TITLE
Nc news 4

### PR DESCRIPTION
**Pull request by HaithemLO**
renamed named it to nc-news 4 to match the sprint

### CODE REVIEW

File: models/model.js [hunk 1]
```diff
@@ -9,7 +10,17 @@ const readTopics = () => {
 };
 
 
+const readArticleById = (article_id) => {
+    return db.query(`SELECT * FROM articles WHERE article_id = $1;`, [article_id])
+      .then(({ rows }) => {
+        if (rows.length === 0) {
+          return Promise.reject({ status: 404, msg: 'Article not found' });
+        }
+        return rows[0];
```
Comment (Pipe-Smoking-Rabbit) [inline models/model.js:19]: Very good model function! Good check for empty rows to send a _Promise.reject_ and good use of the `$` binding 🏆

File: controller/controller.js [hunk 2]
```diff
@@ -11,6 +12,27 @@ const getTopics = (req,res,next) => {
...
+    res.status(200).json(documentation);
+  };
 
+  const getArticleById = (req, res, next) => {
+    const { article_id } = req.params;
+    readArticleById(article_id)
+      .then((article) => {
+        res.status(200).json({ article });
+      })
+      .catch(next);
```
Comment (Pipe-Smoking-Rabbit) [inline controller/controller.js:35]: Great controller! That's a lovely syntax for the catch using `.catch(next)` - but then it doesn't look like anything is happening with that error, how should we handle the error in the _app.js_ ?

**Review by Pipe-Smoking-Rabbit — CHANGES_REQUESTED** [review]
Really good work on this ticket, I've left some feedback to have a look at on the code. One thing that is missing I noticed is any tests for errors, what would we expect if:
- The _article_id_ is a number that can't be found in the data base, such as `99999` ?
- The _article_id_ is the wrong data type, _"/api/articles/banana"_ ?

File: __tests__/app.test.js [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+      .get('/api/invalid')
+      .expect(404);
+  });
+
+  it('should respond with an array of objects that is not empty', () => {
+    return request(app)
+      .get('/api/topics')
+      .expect(200)
+      .then((res)=>{
+        expect(res.body.topics.length).not.toBe(0)
```
Comment (Hard-Luck) [inline __tests__/app.test.js:34]: We can be more specific here, look in the database test data and check what the length should

File: __tests__/app.test.js [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+
+});
+
+
+describe('GET /api/articles/:article_id', () => {
+ 
+  it('should respond with status 404 for an invalid article_id', () => {
+    return request(app)
+      .get('/api/articles/999999999999') // Use a non-existing article_id
+      .expect(404);
```
Comment (Hard-Luck) [inline __tests__/app.test.js:98]: We should also expect a message served back to us with the 404

File: __tests__/app.test.js [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+  });
+
+
+  it('should respond with status 404 for an invalid data type', () => {
+    return request(app)
+      .get('/api/articles/banna') 
+      .expect(404);
```
Comment (Hard-Luck) [inline __tests__/app.test.js:105]: This would be a 400 as the user would have made a bad request

**Review by Hard-Luck — CHANGES_REQUESTED** [review]
just a couple of changes to make

**Review by Keb84 — APPROVED** [review]
Nice work. Merge when ready 🙂

**Pull request merged**